sklego/model_selection.py

```diff
@@ -0,0 +1,44 @@
+from sklearn.model_selection._split import _BaseKFold
+from sklearn.cluster import k_means
+import numpy as np
+import pandas as pd
+import warnings
+
+
+class KMeansFold(_BaseKFold):
+    """
+    K-Means cross validator
+
+    :param n_splits: Number of splits (clusters) to fold on
+    :param kmeans_kwargs: Extra kwargs for k_means apart from n_clusters
+    """
+
+    def __init__(self, n_splits='default', k_means_kwargs=None):
```

Comment:
As with anything sklearn like, I think we should only assume that the `clusterer` has a fit and predict. We could in theory in our fit check whether the `clusterer` is fitted so that you can supply a pretrained clusterer

Reply:
@MBrouns how do you suggest to use the 'check_is_fitted'? I see two problems with this:

- What attribute to check of the clusterer? This might differ per clustering technique.
- How to check if the clusterer is indeed fitted? check_is_fitted does not return a boolean.

Reply:
you could have a `self.custer_model` that is only attached to the object during the `.fit()` step, no?

Reply:
This is how I did it in the `EstimatorTransformer`:

```
    def fit(self, X, y):
        """Fits the estimator"""
        X, y = check_X_y(X, y, estimator=self, dtype=FLOAT_DTYPES)

        self.estimator_ = clone(self.estimator)
        self.estimator_.fit(X, y)
        return self

    def transform(self, X):
        """
        Applies the `predict_func` on the fitted estimator.
        Returns an array of shape `(X.shape[0], )`.
        """
        check_is_fitted(self, 'estimator_')```

Reply:
@koaning True, I agree. However, I need to check **if** I need to call the .fit method.

@MBrouns This method works indeed if you inherit the MetaEstimatorMixin. I am not sure If I want that for this class. Now I am getting the following error: `KlusterFoldValidation(cluster_method=KMeans(algorithm='auto', copy_x=True, init='k-means++', max_iter=300,
    n_clusters=8, n_init=10, n_jobs=None, precompute_distances='auto',
    random_state=None, tol=0.0001, verbose=0),
           random_state=123) is not an estimator instance.
`